rollout/trafficrouting.go

```diff
@@ -17,152 +17,177 @@ import (
 )
 
 // NewTrafficRoutingReconciler identifies return the TrafficRouting Plugin that the rollout wants to modify
-func (c *Controller) NewTrafficRoutingReconciler(roCtx *rolloutContext) (trafficrouting.TrafficRoutingReconciler, error) {
+func (c *Controller) NewTrafficRoutingReconciler(roCtx *rolloutContext) ([]trafficrouting.TrafficRoutingReconciler, error) {
 	rollout := roCtx.rollout
+	// define an empty list of trafficReconcilers to be populated
+	// by the ones declared in the rolloutContext
+	trafficReconcilers := []trafficrouting.TrafficRoutingReconciler{}
+
 	if rollout.Spec.Strategy.Canary.TrafficRouting == nil {
 		return nil, nil
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.Istio != nil {
 		if c.IstioController.VirtualServiceInformer.HasSynced() {
-			return istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, c.IstioController.VirtualServiceLister, c.IstioController.DestinationRuleLister), nil
+			trafficReconcilers = append(trafficReconcilers, istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, c.IstioController.VirtualServiceLister, c.IstioController.DestinationRuleLister))
 		} else {
-			return istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, nil, nil), nil
+			trafficReconcilers = append(trafficReconcilers, istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, nil, nil))
 		}
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.Nginx != nil {
-		return nginx.NewReconciler(nginx.ReconcilerConfig{
+		trafficReconcilers = append(trafficReconcilers, nginx.NewReconciler(nginx.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.kubeclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 			IngressLister:  c.ingressesLister,
-		}), nil
+		}))
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.ALB != nil {
-		return alb.NewReconciler(alb.ReconcilerConfig{
+		alb_reconcilier, err := alb.NewReconciler(alb.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.kubeclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 			IngressLister:  c.ingressesLister,
 		})
+		trafficReconcilers = append(trafficReconcilers, alb_reconcilier)
```

Comment:
If `err != nil` you will be appending a `nil` `alb_reconciler` to the `trafficReconcilers` slice which may cause some negative side effect. It is better to check the error first and just append if everything is fine.

Reply:
I understand your point.
```
		return alb.NewReconciler(alb.ReconcilerConfig{
```
This is the original one, it was returning it in that way before.
Altering this could affect existing tests and behaviour.

I do agree with what you are saying here, there has to be a decision of how we want to action now onwards this multiple trafficRouting scenarios.
Originally, 1 provider could fail and `return nil, err`. So now I believe there are 3 choices:
1. return [valid, nil], err
2. return [nil], err
3. return [valid], err

I feel like number 2 could be a good approach, either you have a valid Rollout or not.. nothing in between.

Reply:
in any case, the errors of the reconcilers are controlled in the `reconcileTrafficRouting()`. No matter what u append here if you have `err != nil` reconcileTrafficRouting will not process any provider and return the `err`

appending whatever is the result, allows further debugging over the []reconcilers. 
I decided to imply the less amount of changes in order to avoid unexpected alterations of the controller behavior. Maintainers can provide a direct guideline on this and I'm fine to alter this.

Reply:
@fblgit you need to check for err in other cases as well. i see error validation partial here. in above conditions like istio/ingress, validation is not done

Reply:
The problem is that the slice is declared as:

```golang
trafficReconcilers := []trafficrouting.TrafficRoutingReconciler{}
```

As `TrafficRoutingReconciler` is an interface, this code could lead in some situation similar to this:

```golang
trafficReconcilers = append(trafficReconcilers, nil)
for _, reconciler := range trafficReconcilers {
    reconciler.SetWeight()
}
```

which will panic with `invalid memory address or nil pointer dereference`

> @fblgit: No matter what u append here if you have err != nil reconcileTrafficRouting will not process any provider and return the err

Regardless if the consumers will verify the error or not, I wouldn't append a `nil` value to a slice to avoid the `nil pointer` error to ever happen.

Reply:
I actually agree with @leoluz . As far as I understand there is no big difference in controller behavior right now but `nil` in the slice might be a surprise for someone in the future. Just to avoid debugging nil dereference error I would not return slice with nil element.

Reply:
change commited.